internal/clients/gitdiff/gitdiff.go

```diff
@@ -0,0 +1,66 @@
+package gitdiff
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+const unknownCommitHash = "0000000000000000000000000000000000000000"
+
+// GitDiff client for getting diffs from the command line
+type GitDiff struct {
+	WorkDir    string
+	BaseBranch string
+	BaseSHA    string
+	Head       string
+}
+
+// GetDiff uses the command line to compute the diff
+func (gd *GitDiff) GetDiff() (string, error) {
+	err := os.Chdir(gd.WorkDir)
+	if err != nil {
+		return "", err
+	}
+
+	var diffCmd *exec.Cmd
+	if gd.BaseBranch != "" {
+		err = exec.Command("git", "fetch", "origin", gd.BaseBranch, "--depth=1").Run()
+		if err != nil {
+			return "", err
+		}
+		diffCmd = exec.Command("git", "diff", fmt.Sprintf("origin/%s", gd.BaseBranch))
+	} else if gd.BaseSHA == "" || gd.BaseSHA == unknownCommitHash {
+		err = exec.Command("git", "fetch", "origin", gd.Head, "--depth=2").Run()
+		if err != nil {
+			return "", err
+		}
+		diffCmd = exec.Command("git", "show", gd.Head, "--format=")
+	} else {
+		err = exec.Command("git", "fetch", "origin", gd.BaseSHA, "--depth=1").Run()
+		if err != nil {
+			return "", err
+		}
+		diffCmd = exec.Command("git", "diff", gd.BaseSHA, gd.Head)
+	}
+	reader, err := diffCmd.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+	err = diffCmd.Start()
+	if err != nil {
+		return "", nil
+	}
+	buf := new(strings.Builder)
+	_, err = io.Copy(buf, reader)
+	if err != nil {
+		return "", err
+	}
+	err = diffCmd.Wait()
```

Comment:
what does wait do?

Reply:
The [`Wait()`](https://golang.org/pkg/os/exec/#Cmd.Wait) waits for the command to finish running. Due to the `Copy()` I do not believe that is waits for long or any time, but it does provide an error if the command fails (returns non zero code).

Reply:
gotcha. Why are we `Start()`ing and then `Wait()`ing here, whereas other places we just call `Run()` (which I assume is synchronous)? Is this the only way we can redirect the output into our buffer for some reason?

Reply:
Yea, this is the only way I can see where we can get a reader for the result. For the others we don't care about the output only about if there is an error so that is why I am using `Run()`